euphonic/spectra.py

```diff
@@ -1297,6 +1297,89 @@ def from_dict(cls: Type[T], d: Dict[str, Any]) -> T:
                    x_tick_labels=d['x_tick_labels'],
                    metadata=d['metadata'])
 
+    def apply_kinematic_constraints(self: T,
+                                    e_i: Quantity = None,
+                                    e_f: Quantity = None,
+                                    angle_range: Tuple[float] = (0, 180.)
+                                    ) -> T:
+        """Set events to NaN which violate energy/momentum limits:
+
+        - Energy transfer greater than e_i
+        - q outside region accessible for given e_i and angle range
+
+        This requires x_data to be in wavevector units and y_data to be energy
+
+        Either e_i or e_f should be set, according to direct/instrument
+        geometry. The other values will be inferred, interpreting y_data as
+        energy transfer.
+
+        """
+
+        momentum2_to_energy = 0.5 * ureg('hbar^2 / neutron_mass').to('meV angstrom^2')
+
+        if (e_i is None) == (e_f is None):
+            raise ValueError("Only one of e_i and e_f should be set. "
+                             "(The other value will be derived from energy "
+                             "transfer).")
+
+        elif e_i is None:   # Indirect geometry: final energy is fixed,
+                            # incident energy range is unlimited
+            e_f = e_f.to('meV')
+            e_i = (self.get_bin_centres(bin_ax='y').to('meV') + e_f)
+        elif e_f is None:   # Direct geometry: incident energy is fixed,
+                            # max energy transfer = e_i
+            e_i = e_i.to('meV')
+            e_f = e_i - self.get_bin_centres(bin_ax='y').to('meV')
+
+        k2_i = (e_i / momentum2_to_energy)
+        k2_f = (e_f / momentum2_to_energy)
+
+        cos_values = np.cos(_get_abs_angle_range(angle_range) * np.pi / 180.)
+
+        # Momentum goes negative where final energy greater than incident
+        # energy; detect this as complex component and set extreme q-bounds to
+        # enforce conservation of energy
+        q_bounds = np.sqrt(k2_i + k2_f
+                           - 2 * cos_values[:, np.newaxis]
+                               * np.sqrt(k2_i * k2_f, dtype=complex)
+                           )
+        q_bounds.magnitude.T[np.any(q_bounds.imag, axis=0)] = [float('Inf'),
+                                                               float('-Inf')]
```

Comment:
There aren't any tests where there are imaginary `q_bounds`, it would be good to have one

Reply:
Good point. Just need one with a higher incident energy.

Reply:
Done! The same case checks for a nasty angle as well.